File: php/sword_offer_25.php
<?php

// 剑指 Offer 25. 合并两个排序的链表

// 输入两个递增排序的链表，合并这两个链表并使新链表中的节点仍然是递增排序的。



//  Definition for a singly-linked list.
  class ListNode {
      public $val = 0;
      public $next = null;
      function __construct($val) { $this->val = $val; }
  }

class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function mergeTwoLists($l1, $l2) {
        $node = new ListNode(0);
        $current = $node;
        while ($l1 != null && $l2 != null){
            if($l1->val > $l2->val){
                $current->next = $l2;
                $l2 = $l2->next;
            }else{
                $current->next = $l1;
                $l1 = $l1->next;
            }
            $current = $current->next;
        }
        if($l1){
            $current->next = $l1;
        }
        if($l2){
            $current->next = $l2;
        }
        return $node->next;
    }

    function listToSlice($l){
        $s = [];
        while ($l != null){
            $s[] = $l->val;
            $l = $l->next;
        }
        return $s;
    }

    function node1(){
        $node = new ListNode(1);
        $node->next = new ListNode(2);
        $node->next->next = new ListNode(4);
        return $node;
    }

    function node2(){
        $node = new ListNode(1);
        $node->next = new ListNode(3);
        $node->next->next = new ListNode(4);
        return $node;
    }
}

$obj = new Solution();
$ret = $obj->mergeTwoLists($obj->node1(),$obj->node2());
$ret = $obj->listToSlice($ret);
var_dump($ret);

File: go/sword_offer_25.go
package main

import (
	"fmt"
)

// 剑指 Offer 25. 合并两个排序的链表

// 输入两个递增排序的链表，合并这两个链表并使新链表中的节点仍然是递增排序的。


  //Definition for singly-linked list.
type ListNode struct {
  Val int
  Next *ListNode
}


func main()  {
	ret := mergeTwoLists(node11(),node22())
	fmt.Println(ret)
	listToSlice(ret)
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	node := &ListNode{
		Val: 0,
	}
	current := node
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			current.Next = l1
			l1 = l1.Next
		}else{
			current.Next = l2
			l2 = l2.Next
		}
		current = current.Next
	}
	// 把没合并完的直接合并
	if l1 != nil {
		current.Next = l1
	}
	if l2 != nil {
		current.Next = l2
	}
	return node.Next
}

func listToSlice(l *ListNode)  {
	var s = make([]int,0)
	for l != nil {
		s = append(s,l.Val)
		l = l.Next
	}
	fmt.Println(s)
}

func node11() *ListNode {
	node := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 4,
			},
		},
	}
	return node
}

func node22() *ListNode {
	node := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 3,
			Next: &ListNode{
				Val: 4,
			},
		},
	}
	return node
}


